### Naive Bayes Assignment

#### Problem Statement
Prepare a classification model using Naive Bayes for salary data 

Dataset --> SalaryData_Test.csv, SalaryData_Train.csv

In [1]:
#Importing the Required Library
import pandas as pd
import numpy as np
from sklearn import preprocessing

#Naive Bayes Library
from sklearn.naive_bayes import MultinomialNB as MB
from sklearn.naive_bayes import GaussianNB as GB

In [2]:
# loading dataset
dataset = pd.read_csv('C:/Users/Akaash/Downloads/SalaryData_Train.csv')
dataset.head()

age          workclass   education  educationno        maritalstatus  \
0   39          State-gov   Bachelors           13        Never-married   
1   50   Self-emp-not-inc   Bachelors           13   Married-civ-spouse   
2   38            Private     HS-grad            9             Divorced   
3   53            Private        11th            7   Married-civ-spouse   
4   28            Private   Bachelors           13   Married-civ-spouse   

           occupation    relationship    race      sex  capitalgain  \
0        Adm-clerical   Not-in-family   White     Male         2174   
1     Exec-managerial         Husband   White     Male            0   
2   Handlers-cleaners   Not-in-family   White     Male            0   
3   Handlers-cleaners         Husband   Black     Male            0   
4      Prof-specialty            Wife   Black   Female            0   

   capitalloss  hoursperweek          native  Salary  
0            0            40   United-States   <=50K  
1            0            13   United-States   <=50K  
2            0            40   United-States   <=50K  
3            0            40   United-States   <=50K  
4            0            40            Cuba   <=50K

In [3]:
#Checking the Dataset for Na Values & Dtypes of Column
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30161 entries, 0 to 30160
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   age            30161 non-null  int64 
 1   workclass      30161 non-null  object
 2   education      30161 non-null  object
 3   educationno    30161 non-null  int64 
 4   maritalstatus  30161 non-null  object
 5   occupation     30161 non-null  object
 6   relationship   30161 non-null  object
 7   race           30161 non-null  object
 8   sex            30161 non-null  object
 9   capitalgain    30161 non-null  int64 
 10  capitalloss    30161 non-null  int64 
 11  hoursperweek   30161 non-null  int64 
 12  native         30161 non-null  object
 13  Salary         30161 non-null  object
dtypes: int64(5), object(9)
memory usage: 3.2+ MB


Inference: There are No NA Values Present,  Also Datasets Has X as well as the Y Variable Categorial, So for Y variable Using Label Encoder then for X Variable will Use Dummies

In [4]:
#Complete SalaryData dataset - applying lable encoder to size_category column - (Y Variable is Categorial)
label_encoder = preprocessing.LabelEncoder()
dataset['Salary'] = label_encoder.fit_transform(dataset['Salary']) 
dataset.head()

age          workclass   education  educationno        maritalstatus  \
0   39          State-gov   Bachelors           13        Never-married   
1   50   Self-emp-not-inc   Bachelors           13   Married-civ-spouse   
2   38            Private     HS-grad            9             Divorced   
3   53            Private        11th            7   Married-civ-spouse   
4   28            Private   Bachelors           13   Married-civ-spouse   

           occupation    relationship    race      sex  capitalgain  \
0        Adm-clerical   Not-in-family   White     Male         2174   
1     Exec-managerial         Husband   White     Male            0   
2   Handlers-cleaners   Not-in-family   White     Male            0   
3   Handlers-cleaners         Husband   Black     Male            0   
4      Prof-specialty            Wife   Black   Female            0   

   capitalloss  hoursperweek          native  Salary  
0            0            40   United-States       0  
1            0            13   United-States       0  
2            0            40   United-States       0  
3            0            40   United-States       0  
4            0            40            Cuba       0

Inference: Label Encoder is Applied on Y Variable So 0 = "<=50" Class , 1 = ">50" Class

In [5]:
#Dummies for X Variable
dataset = pd.get_dummies(dataset)
dataset.head()

age  educationno  capitalgain  capitalloss  hoursperweek  Salary  \
0   39           13         2174            0            40       0   
1   50           13            0            0            13       0   
2   38            9            0            0            40       0   
3   53            7            0            0            40       0   
4   28           13            0            0            40       0   

   workclass_ Federal-gov  workclass_ Local-gov  workclass_ Private  \
0                       0                     0                   0   
1                       0                     0                   0   
2                       0                     0                   1   
3                       0                     0                   1   
4                       0                     0                   1   

   workclass_ Self-emp-inc  ...  native_ Portugal  native_ Puerto-Rico  \
0                        0  ...                 0                    0   
1                        0  ...                 0                    0   
2                        0  ...                 0                    0   
3                        0  ...                 0                    0   
4                        0  ...                 0                    0   

   native_ Scotland  native_ South  native_ Taiwan  native_ Thailand  \
0                 0              0               0                 0   
1                 0              0               0                 0   
2                 0              0               0                 0   
3                 0              0               0                 0   
4                 0              0               0                 0   

   native_ Trinadad&Tobago  native_ United-States  native_ Vietnam  \
0                        0                      1                0   
1                        0                      1                0   
2                        0                      1                0   
3                        0                      1                0   
4                        0                      0                0   

   native_ Yugoslavia  
0                   0  
1                   0  
2                   0  
3                   0  
4                   0  

[5 rows x 103 columns]

Inference: Now the Dataset is Complete, Lets Separate X & Y Variable 

In [6]:
# Dividing our data into X and Y variables 
X = pd.concat([dataset.iloc[:,0:5],dataset.iloc[:,6:]],axis = 1)
Y = dataset.iloc[:,5]

Inference: There are 102 Input Variable & 1 Class Variable

#### Naive Bayes Train Model

In [7]:
# Multinomial Naive Bayes
classifier_mb = MB()
classifier_mb.fit(X,Y)
train_pred = classifier_mb.predict(X)
accuracy_train = np.mean(train_pred==Y)

print("Train Accuracy",accuracy_train)


Train Accuracy 0.7729186698053778


Inference: Train Accuracy is 77.29% by Multinomial Naive Bayes 

In [8]:
# Gaussian Naive Bayes 
classifier_gb = GB()
classifier_gb.fit(X,Y) 
train_pred_g = classifier_gb.predict(X)
accuracy_train_g = np.mean(train_pred_g==Y)

print("Train Accuracy",accuracy_train_g)

Train Accuracy 0.8031563940187659


Inference: Train Accuracy is 80.3% by Gaussian Naive Bayes 

**Testing on the Test Data**

In [9]:
# loading dataset
dataset = pd.read_csv('C:/Users/Akaash/Downloads/SalaryData_Test.csv')
#Complete Forestfires dataset - applying lable encoder to size_category column - (Y Variable is Categorial)
label_encoder = preprocessing.LabelEncoder()
dataset['Salary'] = label_encoder.fit_transform(dataset['Salary']) 
#Dummies for X Variable
dataset = pd.get_dummies(dataset)
# Dividing our data into X and Y variables
X_Test = pd.concat([dataset.iloc[:,0:5],dataset.iloc[:,6:]],axis = 1)
Y_Test = dataset.iloc[:,5]

In [10]:
# Multinomial Naive Bayes (Predicting)
test_pred_m = classifier_mb.predict(X_Test)
accuracy_test_m = np.mean(test_pred_m==Y_Test)
print("Test Accuracy",accuracy_test_m)

Test Accuracy 0.7749667994687915


Inference: Test Accuracy is 77.49% by Multinomial Naive Bayes 

In [11]:
#Gaussian Naive Bayes (Predicting)
test_pred_g = classifier_gb.predict(X_Test)
accuracy_test_g = np.mean(test_pred_g==Y_Test)
print("Test Accuracy",accuracy_test_g)

Test Accuracy 0.8029216467463479


Inference: Test Accuracy is 80.29% by Gaussian Naive Bayes 

##### Inference: So the Gaussian Naive Bayes Model is best with Test Accuracy of 80.29%